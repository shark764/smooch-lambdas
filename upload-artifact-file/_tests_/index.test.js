const axios = require('axios');

jest.mock('axios');
jest.mock('form-data');

const mockGetSecretValue = jest.fn(() => { })
  .mockImplementation(() => ({
    promise: () => ({
      SecretString: JSON.stringify({
        username: 'username',
        password: 'password',
      }),
    }),
  }));

jest.mock('aws-sdk', () => ({
  SecretsManager: jest.fn().mockImplementation(() => ({
    getSecretValue: mockGetSecretValue,
  })),
  SQS: jest.fn().mockImplementation(() => ({
    getQueueUrl: jest.fn().mockImplementation(() => ({ promise: jest.fn().mockImplementation(() => ({ QueueUrl: 'url://testurl' })) })),
    sendMessage: jest.fn().mockImplementation(() => ({ promise: jest.fn() })),
  })),
  config: {
    update: jest.fn(),
  },
}));

const agentEvent = {
  Records: [
    {
      body: JSON.stringify({
        source: 'agent',
        artifactId: '5e31c81640a22c000f5d7f28',
        interactionId: '667802d8-2260-436c-958a-2ee0f71f73f0',
        tenantId: '66d83870-30df-4a3b-8801-59edff162034',
        fileData: {
          mediaUrl: 'http://example.com/image.jpg',
          filename: 'file.jpg',
          contentType: 'image/jpeg',
        },
        message: {
          id: '5e31c81640a22c000f5d7f33',
          text: 'test',
          type: 'agent',
          contentType: 'image',
          file: {
            mediaType: 'image/jpeg',
            mediaUrl: 'https://example.com/image.jpg',
          },
          from: 'Test',
          agentMessageId: '66d83870-30df-4a3b-8801-59edff162034',
          resourceId: '66d83870-30df-4a3b-8801-59edff161111',
          timestamp: 0,
        },
      }),
    },
  ],
};

const customerEvent = {
  Records: [
    {
      body: JSON.stringify({
        source: 'customer',
        artifactId: '5e31c81640a22c000f5d7f28',
        interactionId: '667802d8-2260-436c-958a-2ee0f71f73f0',
        tenantId: '66d83870-30df-4a3b-8801-59edff162034',
        fileData: {
          mediaUrl: 'http://example.com/image1.jpg',
          filename: 'file.jpg',
          contentType: 'image/jpeg',
        },
        message: {
          _id: '5e31c81640a22c000f5d7f33',
          text: 'test',
          type: 'agent',
          contentType: 'image',
          mediaType: 'image/jpeg',
          mediaUrl: 'https://example.com/image1.jpg',
          from: 'Test',
          agentMessageId: '66d83870-30df-4a3b-8801-59edff162034',
          resourceId: '66d83870-30df-4a3b-8801-59edff161111',
          timestamp: 0,
        },
      }),
    },
  ],
};

axios.mockImplementation(() => ({
  data: Buffer.from('test'),
}));

const { handler } = require('../index');

describe('upload-artifact-file', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  describe('Everything is successful', () => {
    afterEach(() => {
      jest.clearAllMocks();
    });
    it('when message comes from smooch', async () => {
      await handler(customerEvent);
      expect(axios.mock.calls[0]).toEqual([{
        method: 'get',
        responseType: 'arraybuffer',
        url: 'https://example.com/image1.jpg',
      }]);
    });

    describe('Walkthrough', () => {
      beforeEach(async () => {
        mockGetSecretValue.mockImplementationOnce(() => ({
          promise: () => ({
            SecretString: JSON.stringify({
              username: 'username',
              password: 'password',
            }),
          }),
        }));
        await handler(agentEvent);
      });
      it('passes in the correct arguments to secretClient.getSecretValue() to get cx credentials', async () => {
        expect(mockGetSecretValue.mock.calls[0]).toEqual(expect.arrayContaining([{
          SecretId: 'us-east-1-dev-smooch-cx',
        }]));
        expect(mockGetSecretValue.mock.calls[0]).toMatchSnapshot();
      });

      it('passes in the correct arguments to axios to retrieve attachment', async () => {
        expect(axios.mock.calls[0]).toMatchSnapshot();
      });

      it('passes in the correct arguments to axios to upload the attachment file', async () => {
        delete axios.mock.calls[1][0].data;
        expect(axios.mock.calls[1]).toMatchSnapshot();
      });
    });
  });

  it('throws an error when there is a problem retrieving cx credentials', async () => {
    try {
      mockGetSecretValue.mockRejectedValueOnce(new Error());
      await handler(customerEvent);
    } catch (error) {
      expect(Promise.reject(new Error('Error retrieving cx credentials'))).rejects.toThrowErrorMatchingSnapshot();
    }
  });

  it('throws an error when there is a problem retrieving attachment', async () => {
    try {
      axios.mockRejectedValueOnce(new Error());
      await handler(customerEvent);
    } catch (error) {
      expect(Promise.reject(new Error('Error retrieving attachment'))).rejects.toThrowErrorMatchingSnapshot();
    }
  });

  it('throws an error when there is a problem uploading artifact file', async () => {
    try {
      axios.mockImplementation((data) => {
        switch (data.url) {
          case 'https://example.com/image.jpg':
            return Promise.resolve({
              data: Buffer.from('test'),
            });
          case 'https://us-east-1-dev-edge.domain/v1/tenants/66d83870-30df-4a3b-8801-59edff162034/interactions/667802d8-2260-436c-958a-2ee0f71f73f0/artifacts/5e31c81640a22c000f5d7f28':
            return Promise.reject(new Error());
          default:
            return Promise.resolve({ data: {} });
        }
      });
      await handler(agentEvent);
    } catch (error) {
      expect(Promise.reject(new Error('Error uploading file to artifact'))).rejects.toThrowErrorMatchingSnapshot();
    }
  });
});
