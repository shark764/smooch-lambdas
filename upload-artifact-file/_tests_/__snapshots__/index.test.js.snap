// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`upload-artifact-file Everything is successful Walkthrough passes in the correct arguments to axios to retrieve attachment 1`] = `
Array [
  Object {
    "method": "get",
    "responseType": "arraybuffer",
    "url": "https://example.com/image.jpg",
  },
]
`;

exports[`upload-artifact-file Everything is successful Walkthrough passes in the correct arguments to axios to upload the attachment file 1`] = `
Array [
  Object {
    "auth": Object {
      "password": "password",
      "username": "username",
    },
    "headers": undefined,
    "maxBodyLength": 52428800,
    "maxContentLength": 52428800,
    "method": "post",
    "url": "https://us-east-1-dev-edge.domain/v1/tenants/66d83870-30df-4a3b-8801-59edff162034/interactions/667802d8-2260-436c-958a-2ee0f71f73f0/artifacts/5e31c81640a22c000f5d7f28",
  },
]
`;

exports[`upload-artifact-file Everything is successful Walkthrough passes in the correct arguments to secretClient.getSecretValue() to get cx credentials 1`] = `
Array [
  Object {
    "SecretId": "us-east-1-dev-smooch-cx",
  },
]
`;

exports[`upload-artifact-file throws an error when there is a problem retrieving attachment 1`] = `"Error retrieving attachment"`;

exports[`upload-artifact-file throws an error when there is a problem retrieving cx credentials 1`] = `"Error retrieving cx credentials"`;

exports[`upload-artifact-file throws an error when there is a problem uploading artifact file 1`] = `"Error uploading file to artifact"`;
